driver/src/main/com/mongodb/reactivestreams/client/internal/GridFSDownloadPublisherImpl.java

```diff
@@ -0,0 +1,214 @@
+        s.onSubscribe(new GridFSDownloadSubscription(s));
+    }
+
+    class GridFSDownloadSubscription implements Subscription {
+        private final Subscriber<? super ByteBuffer> outerSubscriber;
+        private final Object lock = new Object();
+
+        /* protected by `lock` */
+        private GridFSFile gridFSFile;
+        private boolean requestedData;
+        private long sizeRead = 0;
+        private long requested = 0;
+        private int currentBatchSize = 0;
+        private boolean isCompleted = false;
+        private boolean isTerminated = false;
+        /* protected by `lock` */
+
+        public GridFSDownloadSubscription(final Subscriber<? super ByteBuffer> outerSubscriber) {
+            this.outerSubscriber = outerSubscriber;
+        }
+
+        private final Subscriber<GridFSFile> gridFSFileSubscriber = new Subscriber<GridFSFile>() {
+            @Override
+            public void onSubscribe(final Subscription s) {
+                s.request(1);
+            }
+
+            @Override
+            public void onNext(final GridFSFile result) {
+                synchronized (lock) {
+                    gridFSFile = result;
+                }
+            }
+
+            @Override
+            public void onError(final Throwable t) {
+                outerSubscriber.onError(t);
+                terminate();
+            }
+
+            @Override
+            public void onComplete() {
+                synchronized (lock) {
+                    requestedData = false;
+                }
+                requestMoreOrComplete();
+            }
+        };
+
+        class GridFSDownloadStreamSubscriber implements Subscriber<Integer> {
+            private final ByteBuffer byteBuffer;
+
+            GridFSDownloadStreamSubscriber(final ByteBuffer byteBuffer) {
+                this.byteBuffer = byteBuffer;
+            }
+
+            @Override
+            public void onSubscribe(final Subscription s) {
+                s.request(1);
+            }
+
+            @Override
+            public void onNext(final Integer integer) {
+                synchronized (lock) {
+                    sizeRead += integer;
+                }
+            }
+
+            @Override
+            public void onError(final Throwable t) {
+                terminate();
+                outerSubscriber.onError(t);
+            }
+
+            @Override
+            public void onComplete() {
+                if (byteBuffer.remaining() > 0) {
+                    gridFSDownloadStream.read(byteBuffer).subscribe(new GridFSDownloadStreamSubscriber(byteBuffer));
+                } else {
+                    synchronized (lock) {
+                        requestedData = false;
+                        if (sizeRead == gridFSFile.getLength()) {
+                            isCompleted = true;
+                        }
+                    }
+                    byteBuffer.flip();
+                    outerSubscriber.onNext(byteBuffer);
```

Comment:
Read all the data into the byteBuffer - pass it to the main subscribers onNext method.